File: LeetCode/LeetCode Practice Problems/807- Max Increase to Keep City Skyline.cpp
class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        if(grid.size() == 0)
            return 0;
        vector<int> row(grid.size());
        vector<int> column(grid[0].size());
        
        for(int i=0;i<grid.size();i++)
        {
            int maxi = grid[i][0];
            for(int j=0;j<grid[i].size();j++)
                maxi = max(maxi , grid[i][j]);
            
            row[i] = maxi;
        }
        
        for(int j=0;j<grid[0].size();j++)
        {
            int maxi = grid[0][j];
            for(int i=0;i<grid.size();i++)
                maxi = max(maxi , grid[i][j]);
            
            column[j] = maxi;
        }
        
        int ans = 0;
        for(int i=0;i<grid.size();i++)
            for(int j=0;j<grid[0].size();j++)
                ans = ans + min(row[i] , column[j]) - grid[i][j];
        
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/807- Max Increase to Keep City Skyline.java
class Solution {
    public int maxIncreaseKeepingSkyline(int[][] grid) {
        int ans = 0;
        int n = grid.length;
        int rmax[] = new int[n];
        int cmax[] = new int[n];
        
        for(int i=0;i<n;i++) {
            for(int j=0;j<n;j++) {
                rmax[i] = Math.max(rmax[i] , grid[i][j]);
                cmax[i] = Math.max(cmax[i] , grid[j][i]);
            }
        }
        
        for(int i=0;i<n;i++) {
            for(int j=0;j<n;j++) {
                ans += Math.min(rmax[i] , cmax[j]) - grid[i][j];
            }
        }
        
        return ans;
    }
}
